CONTRIBUTING.md

```diff
@@ -0,0 +1,77 @@
+# Contributing guidelines
+
+## Before contributing
+
+Welcome to [TheAlgorithms/Nim](https://github.com/TheAlgorithms/Nim)! Before sending your pull requests, make sure that you __read the whole guidelines__. If you have any doubt on the contributing guide, please feel free to [state it clearly in an issue](https://github.com/TheAlgorithms/Nim/issues/new) or ask the community in [Discord](https://discord.gg/c7MnfGFGa6).
+
+## Contributing
+
+### Contributor
+
+We are very happy that you are considering implementing algorithms and data structures for others! This repository is referenced and used by learners from all over the globe. Being one of our contributors, you agree and confirm that:
+
+- You did your work - no plagiarism allowed
+  - Any plagiarized work will not be merged.
+- Your work will be distributed under [MIT License](LICENSE.md) once your pull request is merged
+- Your submitted work fulfils or mostly fulfills our styles and standards
+
+__New implementation__ is welcome! For example, new solutions for a problem, different representations for a graph data structure or algorithm designs with different complexity but __identical implementation__ of an existing implementation is not allowed. Please check whether the solution is already implemented or not before submitting your pull request.
+
+__Improving comments__ and __writing proper tests__ are also highly welcome.
+
+### Contribution
+
+We appreciate any contribution, from fixing a grammar mistake in a comment to implementing complex algorithms. Please read this section if you are contributing your work.
+
+Please help us keep our issue list small by adding fixes: Add the number of the issue you solved — even if only partially — to the commit message of your pull request. GitHub will use this tag to auto-close the issue when the PR is merged.
+
+#### What is an Algorithm?
+
+An Algorithm is one or more functions (or classes) that:
+* take one or more inputs,
+* perform some internal calculations or data manipulations,
+* return one or more outputs,
+* have minimal side effects (Examples of side effects: `print()`, `plot()`, `read()`, `write()`).
+
+Algorithms should be packaged in a way that would make it easy for readers to put them into larger programs.
+
+Algorithms should:
+* have intuitive class and function names that make their purpose clear to readers
+* use Nim naming conventions and intuitive variable names with correct typing to ease comprehension
+* Prefer `Natural` type to `uint` unless wrapping is mandatory (for e.g. modular algorithms).
+* raise Nim exceptions (`ValueError`, etc.) on erroneous input values
+* add the exceptions raised to the list of side effects
```

Comment:
if we want to make sure that every function accounts for all possible exceptions we can ask to add a `{.push raises: [].}` at the top of every module as adopted in Status style guide: https://status-im.github.io/nim-style-guide/errors.exceptions.html?highlight=raise#exceptions-errorsexceptions. This would force every user to annotate every `func` with `raises` pragma. An example of a the syntax for raise pragma would be useful.

If the procedure is a func and the push raises pragma above is implemented there is no need to track other side effects.

Reply:
Why would one annotate every `func` with `raises` pragma ? I am not sure to understand this one.
func is already a proc annotated with {.noSideEffect.}. Isn't an exception a side effect ?
Why would one want to additionally add a pragma to funcs, shouldn't we call it a proc in this case ?
I do not trust the style guide of Status, it looks quite old.

Isn't this something that we will want to remove, once that Nim 2.0 is out ?

Reply:
I think exceptions are tracked separately from side effects and a func in general can raise side effects. As far as I know status style guide is kept up to date and used. Not saying it should be used as a style guide here, I just happened to learn about that trick there and think is a nice one.

Reply:
I added an example and explained a little bit the usage of this pragma